Implement blocking users

zihadmahiuddin opened the pull request:
Closes #32552 

This PR adds a context menu item to UserPanel for blocking/unblocking a user.
Currently, blocking a friend from the friend list will remove them from the list, not sure if it should match what the website does instead. I attempted that too but wasn't sure what would be the right place to re-fetch the friend list.
Blocked users are also removed from the "Currently Online" list (according to https://github.com/ppy/osu-server-spectator/issues/264).

I was going to implement the related parts in `UserProfileOverlay` but wanted to create the PR first to make sure whether I'm even on the right path with this then look into that one later. Especially since I'm not sure how the block button should be added there.
On the website, it's inside a separate menu but that menu isn't there in-game yet, and just adding a block button with a circled X icon probably wouldn't be clear enough. Also, adding text to the button would make it stick out compared to its neighboring ones. But maybe that's fine for now as it's better than not having a block button at all?

Edit: Marked as draft since I need to know which direction to take for the user profile overlay.

zihadmahiuddin committed:
Implement blocking users from context menu

zihadmahiuddin committed:
Add failing test

zihadmahiuddin committed:
Hide blocked users from currently online

zihadmahiuddin commented:
Actually, maybe it'll be better to do the user profile stuff in a separate PR.

zihadmahiuddin commented:
There is a little problem in the tests [here].(https://github.com/ppy/osu/blob/7ca3a18/osu.Game.Tests/Visual/Online/TestSceneFriendDisplay.cs#L221)
It expects a single `LoadingSpinner` but now `UserPanel` also has one, so that fails.

This diff _fixes_ the tests, but I'm not sure if this is the right way to do it.
```diff
diff --git a/osu.Game.Tests/Visual/Online/TestSceneFriendDisplay.cs b/osu.Game.Tests/Visual/Online/TestSceneFriendDisplay.cs
index 25611cf8d5..52905fe5da 100644
--- a/osu.Game.Tests/Visual/Online/TestSceneFriendDisplay.cs
+++ b/osu.Game.Tests/Visual/Online/TestSceneFriendDisplay.cs
@@ -218,7 +218,7 @@ public void TestLoadFriendsBeforeDisplay()
         }
 
         private void waitForLoad()
-            => AddUntilStep("wait for panels to load", () => this.ChildrenOfType<LoadingSpinner>().Single().State.Value, () => Is.EqualTo(Visibility.Hidden));
+            => AddUntilStep("wait for panels to load", () => this.ChildrenOfType<LoadingSpinner>().First().State.Value, () => Is.EqualTo(Visibility.Hidden));
 
         private void assertVisiblePanelCount<T>(int expectedVisible)
             where T : UserPanel
```

bdach commented on osu.Game/Online/API/IAPIProvider.cs:
I do not feel great at all about `IAPIProvider` sprawling out to include block information like this. To me `LocalUser`, `Friends`, and `Blocks` should probably go live elsewhere at this point (and I am aware the implications the first one especially has on everything). I'm not sure I've ever expressed this out loud but it has struck me as odd for a while that `IAPIProvider` just has a friends list on it.

Not sure this is going to get done because of these implications but I feel like I need to at least voice concern. It shouldn't be `IAPIProvider`'s job to contain all this state imo. At *least* not the friends/blocks stuff. Especially given that we also have yet-to-be-fixed issues like the friends list only updating on launch and when friending someone, so you can open client, then web, then friend someone on web and the client will be blissfully unaware until a reboot, so we'll probably want extra tracking, which will probably bloat `IAPIProvider` even more etc....

peppy replied:
As per my review, I agree with this, but it should probably be done as follow-up cleanup work.

As it stands, I think this PR does what it says on the tin and can probably go forward.

bdach commented on osu.Game/Overlays/Dashboard/CurrentlyOnlineDisplay.cs:
```diff
@@ -104,6 +113,36 @@ protected override void OnFocus(FocusEvent e)
+                    foreach (APIRelation block in e.NewItems.Cast<APIRelation>())
+                    {
+                        int userId = block.TargetID;
+                        removeUserPanel(userId);
```
Is this something that should even be handled client-side at all like this? I'd say no, and based on https://github.com/ppy/osu-server-spectator/issues/264 I'd say I'm not alone in thinking that.

bdach reviewed (commented):
Not feeling too happy with any of this

zihadmahiuddin committed:
Revert "Hide blocked users from currently online"

This reverts commits 7ca3a1895a412092da45c7db10c3c7babec59b40 and fbdea8f99019779e15babf2e41b9dda6da84aa70.

zihadmahiuddin committed:
Fix failing tests

Now that `UserPanel` also has a `LoadingSpinner`, we need to use `.First` instead of `.Single` here.

peppy commented on osu.Game/Users/UserPanel.cs:
```diff
@@ -96,6 +104,7 @@ private void load()
                 Add(background);
 
             Add(CreateLayout());
+            Add(loading = new LoadingLayer(true));
```
I don't think there should be a loading layer for such a rare action as block, especially when it's activated from a menu. Just remove this.

peppy commented on osu.Game/Users/UserPanel.cs:
```diff
@@ -179,9 +192,34 @@ public MenuItem[] ContextMenuItems
             }
         }
 
+        private void blockUser(bool block)
```
We're going to want to be able to block users from other locations in game in the future. I guess this can exist here for now, but we'll eventually want to reconsider.

peppy commented on osu.Game/Users/UserPanel.cs:
```diff
@@ -179,9 +192,34 @@ public MenuItem[] ContextMenuItems
+
+            req.Success += () =>
+            {
+                api.UpdateLocalBlocks();
```
This is super awkward. I've much rather the `Block` / `Unblock` methods are on a global class.

But... it does match the implementation of `UpdateLocalFriends` so I can't really fault the implementation.

peppy committed:
Merge branch 'master' into feat/block-users

peppy committed:
Remove loading layer

peppy commented:
Merging this in because it works and adds functionality we need. Can be refactored at a later point in time alongside similar (already-merged) code.
